### PR TITLE
Make `SinglePlaceholderPluralPatterns` generic and use it in `CurrencyPatternsDataV1`

### PR DESCRIPTION
**Pull request by robertbastian**
Supersedes #5373

### CODE REVIEW

File: provider/source/src/currency/patterns.rs [hunk 1]
```diff
@@ -48,26 +53,23 @@ impl DataProvider<CurrencyPatternsDataV1Marker> for SourceDataProvider {
...
-                        _ => Some((count, pattern?)),
-                    }),
-                ),
+                patterns: PluralPattern::try_new(
+                    patterns.pattern_other.as_deref().ok_or_else(|| {
+                        DataError::custom("Missing patterns")
+                            .with_debug_context(currency_patterns)
+                            .with_debug_context(&req.id)
+                    })?,
+                    patterns.pattern_zero.as_deref(),
```
Comment (younies) [inline provider/source/src/currency/patterns.rs:62]: why these also are not `Pattern`?

File: provider/source/data/debug/currency/patterns@1/ar.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
-  "unit_patterns": {
-    "Other": "{0} {1}"
+  "patterns": {
+    "specials": [],
+    "other": "\u0002\u0005 "
```
Comment (younies) [inline provider/source/data/debug/currency/patterns@1/ar.json:4]: As you can see, the space shifted to the end of the pattern instead in the middle.

For me the PR looks good, let's merge it after the LongCurrencyFormatter PR just to ensure that the test result from this PR are equivalent.

File: provider/source/src/currency/patterns.rs (hunk 1 above)
Comment (robertbastian) [reply provider/source/src/currency/patterns.rs:62]: what do you mean?

File: provider/source/data/debug/currency/patterns@1/ar.json (hunk 2 above)
Comment (robertbastian) [reply provider/source/data/debug/currency/patterns@1/ar.json:4]: This is the string encoding of `icu_pattern`. The escaped values are indices, and the space is the content.

**Review by younies — DISMISSED** [review]

File: provider/source/src/currency/patterns.rs (hunk 1 above)
Comment (younies) [reply provider/source/src/currency/patterns.rs:64]: done offline

File: components/experimental/src/relativetime/provider.rs [hunk 3]
```diff
@@ -101,17 +104,20 @@ pub struct PluralCategoryStr<'data>(pub PluralCategory, pub Cow<'data, str>);
...
 )]
 #[yoke(prove_covariance_manually)]
-pub struct SinglePlaceholderPluralPattern<'data> {
+pub struct PluralPattern<'data, B: PatternBackend<Store = str>> {
     /// Optional patterns for categories other than PluralCategory::Other
     #[cfg_attr(feature = "serde", serde(borrow))]
     pub specials: VarZeroVec<'data, PluralCategoryStrULE>,
     /// The pattern for PluralCategory::Other
     #[cfg_attr(feature = "serde", serde(borrow))]
     pub other: Cow<'data, str>,
```
Comment (sffc) [inline components/experimental/src/relativetime/provider.rs:113]: Nit: The documentation should ideally say that the `str` is the serialized store string, not the human-readable pattern string

**Review by sffc — COMMENTED** [review]
Praise: Good work, consistent with the design of `Pattern`

**Comment by robertbastian** [comment]
https://github.com/unicode-org/icu4x/issues/5369

**Pull request merged**